airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py

```diff
@@ -192,12 +195,12 @@ def stream_slices(self) -> Iterable[StreamSlice]:
                             )
                         )
 
-                # A final parent state update and yield of records is needed, so we don't skip records for the final parent slice
+                # A final yield of records and parent state update is needed, so we don't skip records for the final parent slice
+                yield from stream_slices_for_parent
```

Comment:
This was a subtle but necessary change. We should emit the parent records first before we update the parent state. Otherwise, if we update state first, we may emit a checkpoint state message of the final state of the parent even though we haven't fully processed and iterated through all of the parents.

In `core.py:read()` we should always emit a final state message at the end, so that should capture the final updated parent state. However, even if it weren't, not checkpointing up to the end for an incremental would just mean a few more duplicates which is better than missing records